DOCSP-35966 query builder (#2790)
jenssegers/laravel-mongodb: docs/query-builder.txt

   :values: tutorial

.. meta::
   :keywords: code example, aggregation

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to use the {+odm-short+} extension of
the Laravel query builder to work with a MongoDB database. The query builder
lets you write queries for any supported database by using the same fluent

> interface and syntax.

Reviewer: Q: What does "same" mean in this context?

Author: I think I had "a common" previously to represent the "fluent interface and syntax" which might be more clear than "the same". Let me know if one is more clear than the other, or how you might otherwise word it.

Reviewer: maybe you can just add "... the same ... as the Laravel query builder"?

Author: Thanks for that perspective!

I wanted to describe the query builder rather than comparing it with something else, but it seems like using noun phrases (e.g. https://github.com/mongodb/laravel-mongodb/pull/2790#discussion_r1538001671) can cause some confusion.
In this sentence, the items being compared are syntax you use to access each of the supported databases rather than something with the query builder (given it's describing the query builder).

I'll spend a few minutes to try to reorder the sentence in a way that avoids connecting noun phrases to restrictive clauses.

.. note::

   {+odm-short+} extends Laravel's query builder and Eloquent ORM, which can
   run similar database operations. To learn more about retrieving documents
   by using Eloquent models, see :ref:`laravel-fundamentals-retrieve`.

Laravel provides a **facade** to access the query builder class ``DB``, which
lets you perform database operations. Facades are static interfaces to
classes that make the syntax more concise, avoid runtime errors, and improve

> testability.

Reviewer: nit: ambiguous identifiers - do the classes made the syntax more concise, etc or the facades

{+odm-short+} aliases the ``DB`` method ``table()`` as the ``collection()``
method. Chain methods to specify the command and any constraints. Then, chain

> the ``get()`` method at the end to run them on the MongoDB collection. The

Reviewer: ```suggestion
the ``get()`` method at the end to run the command on the MongoDB collection. The
```

Author: Thanks for spotting this! I'll reword since "them" refers to the unspecified number of chained methods.

following example shows the syntax of a query builder call:

.. code-block:: php

   DB::collection('<collection name>')
       // chain methods by using the "->" object operator
       ->get();

> This guide shows examples of the following types of query builder operations:

Reviewer: ```suggestion
This guide provides examples for the following types of query builder operations:
```

Author: I considered the suggestion while, but think that "for" might not be the correct preposition. I think "for" would imply that the examples are being delivered to someone or something. I'll use "provides" and pair it with "of".

- :ref:`laravel-retrieve-query-builder`
- :ref:`laravel-modify-results-query-builder`
- :ref:`laravel-mongodb-read-query-builder`
- :ref:`laravel-mongodb-write-query-builder`

Before You Get Started
----------------------

To run the code examples in this guide, complete the
:ref:`Quick Start <laravel-quick-start>` tutorial to configure a web
application, load sample datasets into your MongoDB deployment, and
run the example code from a controller method.

To perform read and write operations by using the query builder, import the

> ``Illuminate\Support\Facades\DB`` facade and compose your query.

Reviewer: Q: would it be helpful to show a code block with the import statement?

Author: Good question! Based on feedback in a prior PR from @GromNaN, the developer would be using an IDE which would auto-suggest the imports. In this case, although it might get auto-suggested, I think it's helpful to mention the import package to help the reader know where it came from.

           ...
         },
         { "title": "The Real Miyagi",
           "year": 2015,
           "runtime": 90,
           "imdb": {
             "rating": 9.3,
             "votes": 41,
             "id": 2313306
           },
           "plot": "The life of the greatest karate master of a generation.",
            ...
         }
      ]

.. _laravel-query-builder-logical-operations:

Logical Conditional Operations
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The examples in this section show the query builder syntax you
can use to perform the following logical conditional operations:

- :ref:`Logical OR to match one or more conditions <laravel-query-builder-logical-or>`
- :ref:`Logical AND to match all conditions <laravel-query-builder-logical-and>`
- :ref:`Logical NOT to match the negation of the condition <laravel-query-builder-logical-not>`
- :ref:`Nested logical operator groups <laravel-query-builder-logical-nested>`

.. _laravel-query-builder-logical-or:

Logical OR Example
^^^^^^^^^^^^^^^^^^

The following example shows how to chain the ``orWhere()``
query builder method to retrieve documents from the
``movies`` collection that either match the ``year``

> value of ``1955`` or match the title "``Back to the Future``":

Reviewer: ```suggestion
value of ``1955`` or match the title ``"Back to the Future"``:
```

.. literalinclude:: /includes/query-builder/MovieController.php
   :language: php
   :dedent:
   :start-after: begin query orWhere
   :end-before: end query orWhere

> 

Reviewer: Q: is there no output for these examples, or did you purposely exclude the output box for brevity?

Author: Good question! 
Yes, I purposely excluded it. In earlier drafts, I had started including the output for everything. However, I consciously decided to exclude output of the examples in which the output wouldn't contribute to building understanding since they would either have to know details about the dataset and/or the result is very similar to the description of the query.

.. _laravel-query-builder-logical-and:

Logical AND Example
^^^^^^^^^^^^^^^^^^^

The following example shows how to chain the ``where()``
query builder method to retrieve documents from the
``movies`` collection that match both an ``imdb.rating``

> value greater than "8.5" and a ``year`` value of less than

Reviewer: ```suggestion
value greater than ``8.5`` and a ``year`` value of less than
```

``1940``:

.. literalinclude:: /includes/query-builder/MovieController.php
   :language: php
   :dedent:
   :start-after: begin query andWhere
   :end-before: end query andWhere

.. _laravel-query-builder-logical-not:

Logical NOT Example
^^^^^^^^^^^^^^^^^^^

The following example shows how to chain the ``where()``
query builder method to retrieve documents from the
``movies`` collection that match both an ``imdb.rating``
value greater than ``8.5`` and a ``year`` value of less than
``1940``:

> 

Reviewer: I: this description does not match the code example
S: update the description to match or update the code to match

Might want to double check all instances - I will also be looking

Author: Good catch! Copy-pasta from the previous example. Will check the rest again!

.. literalinclude:: /includes/query-builder/MovieController.php
   :language: php
   :dedent:
   :start-after: begin query whereNot
   :end-before: end query whereNot

.. _laravel-query-builder-logical-nested:

Nested Logical Operator Group Example
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The following example shows how to chain the ``where()``
query builder method to retrieve documents from the
``movies`` collection that match both of the following
conditions:

- ``imdb.rating`` value is greater than ``8.5``
- ``year`` value is either ``1986`` or ``1996``

.. literalinclude:: /includes/query-builder/MovieController.php
   :language: php
   :dedent:
   :start-after: begin query nestedLogical
   :end-before: end query nestedLogical

.. _laravel-query-builder-range-type:

Ranges and Type Checks
~~~~~~~~~~~~~~~~~~~~~~

The examples in this section show the query builder syntax you

> can use to perform to match values by using the following range

Reviewer: ```suggestion
can use to match values by using the following range
```

queries and type check operations:

- :ref:`Values within a numerical range <laravel-query-builder-wherebetween>`
- :ref:`Null or missing values <laravel-query-builder-null>`
- :ref:`One or more values of a set <laravel-query-builder-wherein>`
- :ref:`Match dates <laravel-query-builder-wheredate>`
- :ref:`Match a text pattern <laravel-query-builder-pattern>`

.. _laravel-query-builder-wherebetween:

Numerical Range Example
^^^^^^^^^^^^^^^^^^^^^^^

The following example shows how to use the ``whereBetween()``
query builder method to retrieve documents from the
``movies`` collection that contain an ``imdb.rating`` value
between ``9`` and ``9.5``:

.. io-code-block::
   :copyable: true

   .. input:: /includes/query-builder/MovieController.php
      :language: php
      :dedent:
      :start-after: begin query whereBetween
      :end-before: end query whereBetween

   .. output::
      :language: json
      :visible: false

      [
         { "title" "The Godfather", "imdb": { "rating": 9.2, "votes": 1038358, "id": 68646 }, ... },
         { "title": "Hollywood", "imdb": { "rating": 9.1, "votes": 511,"id": 80230 }, ... },

>          { "title": "Cosmos", "imdb": { "rating": 9.3, "votes": 17174, "id": 81846 }, ... },

Reviewer: I: the text highlighting is a little off here. Is this a known issue or is there a different highlighting language we can use?

Author: Thanks for catching that. I'll use "none" instead of JSON. I think it probably fails on the "..." as that's not valid JSON.

         ...
      ]

.. _laravel-query-builder-null:

Null or Missing Values Example
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The following example shows how to use the ``whereNull()``
query builder method to retrieve documents from the
``movies`` collection that omit a ``runtime`` value
or field.

.. literalinclude:: /includes/query-builder/MovieController.php
   :language: php
   :dedent:
   :start-after: begin query whereNull

>    :end-before: end query whereNull

Reviewer: I: no output

Author: Thanks for raising this. I purposely omitted this given that it's demonstrating omission and it probably raises questions as to why there's a missing field. Additional reasons for omitting output 
 listed here: https://github.com/mongodb/laravel-mongodb/pull/2790#discussion_r1538273062

.. seealso::

>    The ``exists()`` MongoDB operation performs the same query.

Reviewer: Q: wouldnt the same function be negating the `exists` method?

Author: Yes, that's correct. Good catch, will update.

wildcard characters:

- ``%`` which matches zero or more characters
- ``_`` which matches a single character

.. io-code-block::
   :copyable: true

   .. input:: /includes/query-builder/MovieController.php
      :language: php
      :dedent:
      :start-after: begin query like
      :end-before: end query like

   .. output::
      :language: json
      :visible: false

      [
         { "title": "Kiss of the Spider Woman", ... },
         { "title": "Spider-Man", ... },
         { "title": "Spider-Man 2", ...},
         ...
      ]

.. _laravel-query-builder-distinct:

Retrieve Distinct Values
~~~~~~~~~~~~~~~~~~~~~~~~

The following example shows how to use the ``distinct()``
query builder method to retrieve all the different values
of the ``year`` field for documents in the ``movies`` collections.

.. literalinclude:: /includes/query-builder/MovieController.php
   :language: php
   :dedent:
   :start-after: begin query distinct
   :end-before: end query distinct

.. _laravel-query-builder-aggregations:

Aggregations
~~~~~~~~~~~~

The examples in this section show the query builder syntax you
can use to perform aggregations, a set of calculations on the

> query results, to return the following information:

Reviewer: S: break up sentences

```suggestion
The examples in this section show the query builder syntax you
can use to perform aggregations. Aggregations are a set of calculations on the
query results that can help you to return the following information about your data:
```

Author: Thanks for the suggestion. I'll break up the sentences to make it more clear.

- :ref:`Results grouped by common field values <laravel-query-builder-aggregation-groupby>`

> - :ref:`Count the of results <laravel-query-builder-aggregation-count>`

Reviewer: ```suggestion
- :ref:`Count the number of results <laravel-query-builder-aggregation-count>`
```

Author: Good catch!

This example chains the following operations to match documents from the
``movies`` collection that contain a ``rated`` value of ``G`` and include the
``title`` field of one movie for each distinct ``runtime`` value:

- Match only documents that contain a ``rated`` field value of "``G``" by
  using the ``where()`` method
- Group data by the distinct values of the ``runtime`` field, which is
  assigned the ``_id`` field, by using the ``groupBy()`` method
- Sort the groups by the ``runtime`` field by using the ``orderBy()`` method
- Return ``title`` data from the last document in the grouped result by
  specifying it in the ``get()`` method

.. tip::

   The ``groupBy()`` method calls the MongoDB ``$group`` aggregation operator
   and ``$last`` accumulator operator. To learn more about these operators, see
   :manual:`$group (aggregation) </reference/operator/aggregation/group/>`
   in the {+server-docs-name+}.

.. io-code-block::
   :copyable: true

   .. input:: /includes/query-builder/MovieController.php
      :language: php
      :dedent:
      :start-after: begin query groupBy
      :end-before: end query groupBy

   .. output::
      :language: json

>       :visible: false

Reviewer: S: this example would be more illustrative if it produced a result where there were multiple titles per runtime.

Author: Thanks for the suggestion! I think it doesn't work that way since it applies the $last aggregation operator to fields that are not part of the aggregated _id as mentioned in the tip admonition, but perhaps @GromNaN can suggest a solution?

Average Value of a Field Example
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The following example shows how to use the ``avg()``
query builder method to return the numerical average, or
arithmetic mean, of the ``imdb.rating`` values from
the entire ``movies`` collection.

.. literalinclude:: /includes/query-builder/MovieController.php
   :language: php
   :dedent:
   :start-after: begin aggregation avg
   :end-before: end aggregation avg

.. _laravel-query-builder-aggregation-sum:

Summed Value of a Field Example
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The following example shows how to use the ``sum()``
query builder method to return the numerical total of
the ``imdb.votes`` values from the entire ``movies``
collection:

.. literalinclude:: /includes/query-builder/MovieController.php
   :language: php
   :dedent:
   :start-after: begin aggregation sum
   :end-before: end aggregation sum

.. _laravel-query-builder-aggregate-matched:

Aggregate Matched Results Example
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The following example shows how to use the ``sum()``
query builder method to return the numerical total of
the ``imdb.votes`` values from movies released in

> after the year "2000":

Reviewer: I: description does not match the example

Author: Good catch, thanks!

> The following example shows how to use the ``sum()``
> query builder method to return the numerical total of
> the ``imdb.votes`` values from movies released in
> after the year "2000":

Reviewer: ```suggestion
The following example shows how to use the ``sum()``
query builder method to return the numerical total of
the ``imdb.votes`` values from movies released in
after the year ``2000``:
```

- :ref:`Paginate the results <laravel-query-builder-paginate>`

.. _laravel-query-builder-orderby:

Order Results Example
~~~~~~~~~~~~~~~~~~~~~

The following example shows how to use the ``orderBy()``
query builder method to arrange the results that match
the filter specified in the ``title`` field by the
``imdb.rating`` value in descending order:

.. io-code-block::
   :copyable: true

   .. input:: /includes/query-builder/MovieController.php
      :language: php
      :dedent:
      :start-after: begin query orderBy
      :end-before: end query orderBy

   .. output::
      :language: json
      :visible: false

      [
        { "title": "Back to the Future", "imdb": { "rating":8.5,"votes":636511,"id":88763 }, ... },
        { "title": "Back to the Future Part II", "imdb": { "rating":7.8,"votes":292539,"id":96874 }, ... },
        { "title": "Back to the Future Part III", "imdb": {"rating":7.4,"votes":242390,"id":99088 }, ... },
        ...
      ]

.. _laravel-query-builder-skip:

Omit a Specified Number of Results Example
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The following example shows how to use the ``skip()`` query builder method to

> omit the first "4" results that match the filter specified in the ``title``

Reviewer: ```suggestion
omit the first 4 results that match the filter specified in the ``title``
```

Author: Thanks for the suggestion! I'll use "four" as the quantifier since it's less than 10.

> field, sorted by the ``year`` value in descending order:

Reviewer: I; match example
```suggestion
field, sorted by the ``year`` value in ascending order:
```

Author: Good catch!

documents that contain all the specified fields:

.. literalinclude:: /includes/query-builder/MovieController.php
   :language: php
   :dedent:
   :start-after: begin query all
   :end-before: end query all

To learn more about this query operator, see
:manual:`$all </reference/operator/query/all/>`
in the {+server-docs-name+}.

.. _laravel-query-builder-size:

Match Array Size Example
~~~~~~~~~~~~~~~~~~~~~~~~

The following example shows how to pass the ``size``
query operator with the ``where()`` query builder
method to match documents that contain a ``directors``
field that contains an array of exactly five elements:

.. literalinclude:: /includes/query-builder/MovieController.php
   :language: php
   :dedent:
   :start-after: begin query size
   :end-before: end query size

To learn more about this query operator, see
:manual:`$size </reference/operator/query/size/>`
in the {+server-docs-name+}.

.. _laravel-query-builder-type:

Match Data Type Example
~~~~~~~~~~~~~~~~~~~~~~~

The following example shows how to pass the ``type``
query operator with the ``where()`` query builder
method to match documents that contain a ``directors``

> field that contains an array of exactly five elements:

Reviewer: I: description and example do not demonstrate the functionality described in this section for type

Author: Good catch, thanks!

To learn more about this query operator, see
:manual:`$mod </reference/operator/query/mod/>`
in the {+server-docs-name+}.

.. _laravel-query-builder-regex:

Match a Regular Expression
~~~~~~~~~~~~~~~~~~~~~~~~~~

The following example shows how to pass the ``REGEX``
query operator with the ``where()`` query builder
method to match documents that contain a ``title``
field that matches the specified regular expression:

.. literalinclude:: /includes/query-builder/MovieController.php
   :language: php
   :dedent:
   :start-after: begin query whereRegex
   :end-before: end query whereRegex

To learn more about regular expression queries in MongoDB, see
:manual:`$regex </reference/operator/query/regex/>`
in the {+server-docs-name+}.

.. _laravel-query-builder-whereRaw:

Run MongoDB Query API Operations Example
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The following example shows how to use the ``whereRaw()``
query builder method to run a query operation written by
using the MongoDB Query API syntax:

.. literalinclude:: /includes/query-builder/MovieController.php
   :language: php
   :dedent:
   :start-after: begin query raw
   :end-before: end query raw

> The following code shows the equivalent MongoDB query:

Reviewer: ```suggestion
The following code shows the equivalent MongoDB Shell query:
```

Author: Thanks for the suggestion! I think it could be unexpected to introduce MongoDB Shell in passing, so I'll re-iterate that it is MongoDB Query API syntax.

The examples in this section show the query builder syntax you
can use to perform geospatial queries on GeoJSON or coordinate
pair data to retrieve the following types of locations:

- :ref:`Near a position <laravel-query-builder-geospatial-near>`
- :ref:`Within a the boundary of a GeoJSON object <laravel-query-builder-geospatial-geoWithin>`
- :ref:`Intersecting a GeoJSON object <laravel-query-builder-geospatial-geoIntersects>`
- :ref:`Proximity data for nearby matches <laravel-query-builder-geospatial-geoNear>`

.. important::

   To perform GeoJSON queries in MongoDB, you must create either
   ``2d`` or ``2dsphere`` index on the collection. To learn more
   about these indexes, see :manual:`Geospatial Indexes </core/indexes/index-types/index-geospatial/>`
   in the {+server-docs-name+}.

To learn more about GeoJSON objects that MongoDB supports,
see :manual:`GeoJSON Objects </reference/geojson/>`
in the {+server-docs-name+}.

.. _laravel-query-builder-geospatial-near:

Near a Position Example
~~~~~~~~~~~~~~~~~~~~~~~

The following example shows how to use the ``near`` query operator
with the ``where()`` query builder method to match documents that contain the

> field ``random_review``:

Reviewer: I; description does not match

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The following example shows how to use the ``geoInstersects``
query operator with the ``where()`` query builder method to
match documents that contain a location that intersects with
the specified ``LineString`` GeoJSON object:

.. literalinclude:: /includes/query-builder/MovieController.php
   :language: php
   :dedent:
   :start-after: begin query geoIntersects
   :end-before: end query geoIntersects

.. _laravel-query-builder-geospatial-geoNear:

Proximity Data for Nearby Matches Example
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The following example shows how to use the ``geoNear`` aggregation operator
with the ``raw()`` query builder method to perform an aggregation that returns
metadata, such as proximity information for each match:

.. literalinclude:: /includes/query-builder/MovieController.php
   :language: php
   :dedent:
   :start-after: begin query geoNear
   :end-before: end query geoNear

To learn more about this aggregation operator, see
:manual:`$geoNear operator </reference/operator/aggregation/geoNear/>`
in the {+server-docs-name+}.

.. _laravel-mongodb-write-query-builder:

MongoDB Write Operations
------------------------

> 

Reviewer: S: add introduction to section list

- :ref:`Increment a numerical value <laravel-mongodb-query-builder-increment>`
- :ref:`Decrement a numerical value <laravel-mongodb-query-builder-decrement>`
- :ref:`Add an array element <laravel-mongodb-query-builder-push>`
- :ref:`Remove a value from an array <laravel-mongodb-query-builder-pull>`
- :ref:`Remove a field from a document <laravel-mongodb-query-builder-unset>`

.. _laravel-mongodb-query-builder-upsert:

Upsert a Document Example
~~~~~~~~~~~~~~~~~~~~~~~~~

The following example shows how to use the ``update()`` query builder method
and ``upsert`` option to update the matching document or insert one with the
specified data if it does not exist. When setting the ``upsert`` option to
``true``, if the document does not exist, the command inserts both the data
and the ``title`` field and value specified in the ``where()`` query operation:

.. literalinclude:: /includes/query-builder/MovieController.php
   :language: php
   :dedent:
   :start-after: begin upsert
   :end-before: end upsert

.. _laravel-mongodb-query-builder-increment:

Increment a Numerical Value Example
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The following example shows how to use the ``increment()``

> query builder method to add "3000" to the value of

Reviewer: ```suggestion
query builder method to add ``3000`` to the value of
```

the ``imdb.votes`` field in the matched document:

.. literalinclude:: /includes/query-builder/MovieController.php
   :language: php
   :dedent:
   :start-after: begin increment
   :end-before: end increment

.. _laravel-mongodb-query-builder-decrement:

Decrement a Numerical Value Example
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The following example shows how to use the ``decrement()`` query builder
method to subtract ``0.2`` from the value of the ``imdb.rating`` field in the
matched document:

.. literalinclude:: /includes/query-builder/MovieController.php
   :language: php
   :dedent:
   :start-after: begin decrement
   :end-before: end decrement

.. _laravel-mongodb-query-builder-push:

Add an Array Element Example
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The following example shows how to use the ``push()`` query builder method to

> add "``Gary Cole``" to the ``cast`` array field in the matched document:

Reviewer: ```suggestion
add ``"Gary Cole"`` to the ``cast`` array field in the matched document:
```

.. literalinclude:: /includes/query-builder/MovieController.php
   :language: php
   :dedent:
   :start-after: begin push
   :end-before: end push

.. _laravel-mongodb-query-builder-pull:

Remove an Array Element Example
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The following example shows how to use the ``pull()`` query builder method

> to remove the "``Adventure``" value from the ``genres`` field from the document

Reviewer: ```suggestion
to remove the ``"Adventure"`` value from the ``genres`` field from the document
```

interface and syntax.

.. note::

   {+odm-short+} extends Laravel's query builder and Eloquent ORM, which can
   run similar database operations. To learn more about retrieving documents
   by using Eloquent models, see :ref:`laravel-fundamentals-retrieve`.

Laravel provides a **facade** to access the query builder class ``DB``, which
lets you perform database operations. Facades, which are static interfaces to
classes, make the syntax more concise, avoid runtime errors, and improve
testability.

{+odm-short+} aliases the ``DB`` method ``table()`` as the ``collection()``
method. Chain methods to specify commands and any constraints. Then, chain
the ``get()`` method at the end to run the methods and retrieve the results.
The following example shows the syntax of a query builder call:

.. code-block:: php

   DB::collection('<collection name>')
       // chain methods by using the "->" object operator
       ->get();

This guide provides examples of the following types of query builder operations:

- :ref:`laravel-retrieve-query-builder`
- :ref:`laravel-modify-results-query-builder`
- :ref:`laravel-mongodb-read-query-builder`

> - :ref:`laravel-mongodb-write-query-builder`

Reviewer: S: Make the titles of the subsections all in the same form (action oriented)

``Illuminate\Support\Facades\DB`` facade and compose your query.

.. _laravel-retrieve-query-builder:

Retrieve Matching Documents
---------------------------

This section includes query builder examples for read operations in the
following operator categories:

- :ref:`Where method <laravel-query-builder-where-example>`
- :ref:`Logical conditionals <laravel-query-builder-logical-operations>`
- :ref:`Ranges and type checks <laravel-query-builder-range-type>`
- :ref:`Pattern searches <laravel-query-builder-pattern>`
- :ref:`Retrieve distinct values <laravel-query-builder-distinct>`
- :ref:`Aggregations <laravel-query-builder-aggregations>`

.. _laravel-query-builder-where-example:

Where Method Example
~~~~~~~~~~~~~~~~~~~~

The following example shows how to use the ``where()`` query
builder method to retrieve documents from the ``movies`` collection
that contain an ``imdb.votes`` field value of ``350``. Click the
:guilabel:`{+code-output-label+}` button to see the results returned
by the query:

.. io-code-block::
   :copyable: true

   .. input:: /includes/query-builder/QueryBuilderTest.php
      :language: php
      :dedent:
      :start-after: begin query where

>       :end-before: end query where

Reviewer: I: description does not match example
imdb.votes field value of 350
where('imdb.rating', 9.3)

           ...
         },
         { "title": "The Real Miyagi",
           "year": 2015,
           "runtime": 90,
           "imdb": {
             "rating": 9.3,
             "votes": 41,
             "id": 2313306
           },
           "plot": "The life of the greatest karate master of a generation.",
            ...
         }
      ]

.. _laravel-query-builder-logical-operations:

Logical Conditional Operations
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The examples in this section show the query builder syntax you
can use to perform the following logical conditional operations:

- :ref:`Logical OR to match one or more conditions <laravel-query-builder-logical-or>`
- :ref:`Logical AND to match all conditions <laravel-query-builder-logical-and>`
- :ref:`Logical NOT to match the negation of the condition <laravel-query-builder-logical-not>`
- :ref:`Nested logical operator groups <laravel-query-builder-logical-nested>`

.. _laravel-query-builder-logical-or:

Logical OR Example
^^^^^^^^^^^^^^^^^^

The following example shows how to chain the ``orWhere()``
query builder method to retrieve documents from the
``movies`` collection that either match the ``year``

> value of ``1955`` or match the title ``"Back to the Future"``:

Reviewer: ```suggestion
value of ``1955`` or match the ``title`` value ``"Back to the Future"``:
```

.. _laravel-query-builder-wherebetween:

Numerical Range Example
^^^^^^^^^^^^^^^^^^^^^^^

The following example shows how to use the ``whereBetween()``
query builder method to retrieve documents from the
``movies`` collection that contain an ``imdb.rating`` value
between ``9`` and ``9.5``:

.. io-code-block::
   :copyable: true

   .. input:: /includes/query-builder/QueryBuilderTest.php
      :language: php
      :dedent:
      :start-after: begin query whereBetween
      :end-before: end query whereBetween

   .. output::
      :language: none
      :visible: false

      [
         { "title" "The Godfather", "imdb": { "rating": 9.2, "votes": 1038358, "id": 68646 }, ... },
         { "title": "Hollywood", "imdb": { "rating": 9.1, "votes": 511,"id": 80230 }, ... },
         { "title": "Cosmos", "imdb": { "rating": 9.3, "votes": 17174, "id": 81846 }, ... },
         ...
      ]

.. _laravel-query-builder-null:

Null or Missing Values Example
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The following example shows how to use the ``whereNull()``
query builder method to retrieve documents from the
``movies`` collection that omit a ``runtime`` value

> or field.

Reviewer: ```suggestion
or field:
```

.. literalinclude:: /includes/query-builder/QueryBuilderTest.php
   :language: php
   :dedent:
   :start-after: begin query whereNull
   :end-before: end query whereNull

.. _laravel-query-builder-wherein:

One or More Values of a Set Example
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The following example shows how to use the ``whereIn()``
query builder method to retrieve documents from the
``movies`` collection that match at least one of the

> ``title`` values in the specified set.

Reviewer: ```suggestion
``title`` values in the specified set:
```

   .. output::
      :language: json
      :visible: false

      [
         { "title": "Toy Story", "year": 1995, "runtime": 81, ... },
         { "title": "Johnny English", "year": 2003, "runtime": 87, ... },
         { "title": "Shrek 2", "year" 2004, "runtime": 93, ... },
         ...
      ]

.. _laravel-query-builder-wheredate:

Match Dates Example
^^^^^^^^^^^^^^^^^^^

The following example shows how to use the ``whereDate()``
query builder method to retrieve documents from the
``movies`` collection that match the specified date of
``2010-1-15`` in the ``released`` field:

.. literalinclude:: /includes/query-builder/QueryBuilderTest.php
   :language: php
   :dedent:
   :start-after: begin query whereDate
   :end-before: end query whereDate

.. _laravel-query-builder-pattern:

Text Pattern Match Example
~~~~~~~~~~~~~~~~~~~~~~~~~~

The following example shows how to use the ``like`` query operator
with the ``where()`` query builder method to retrieve documents from the
``movies`` collection by using a specified text pattern.

Text patterns can contain text mixed with the following
wildcard characters:

- ``%`` which matches zero or more characters

> - ``_`` which matches a single character

Reviewer: Q: if the `_` matches a single character, why is Spider Woman in the results?

Author: > Q: if the `_` matches a single character, why is Spider Woman in the results?

Good question -- whitespace is a character in the context of SQL queries. The [Laravel documentation](https://laravel.com/docs/11.x/queries#where-any-all-clauses) omits any explanation of the wildcard characters that they use, but I thought it would be good to reiterate the SQL LIKE syntax.
Let me know if you think this needs further explanation.

.. literalinclude:: /includes/query-builder/QueryBuilderTest.php
   :language: php
   :dedent:
   :start-after: begin query size
   :end-before: end query size

To learn more about this query operator, see
:manual:`$size </reference/operator/query/size/>`
in the {+server-docs-name+}.

.. _laravel-query-builder-type:

Match Data Type Example
~~~~~~~~~~~~~~~~~~~~~~~

The following example shows how to pass the ``type``
query operator with the ``where()`` query builder
method to match documents that contain a type ``4`` value,
which corresponds to an array data type, in the
``released`` field.

.. literalinclude:: /includes/query-builder/QueryBuilderTest.php
   :language: php
   :dedent:
   :start-after: begin query type
   :end-before: end query type

To learn more about the type codes and query operator, see
:manual:`$type </reference/operator/query/type/>`
in the {+server-docs-name+}.

.. _laravel-query-builder-mod:

Match a Value Computed with Modulo Example
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The following example shows how to pass the ``mod``
query operator with the ``where()`` query builder
method to match documents by using the expression

> ``"year % 2 == 0"``, which matches even values for

Reviewer: S: remove quotes from the expression 
```suggestion
``year % 2 == 0``, which matches even values for
```

can use to perform geospatial queries on GeoJSON or coordinate
pair data to retrieve the following types of locations:

- :ref:`Near a position <laravel-query-builder-geospatial-near>`
- :ref:`Within a the boundary of a GeoJSON object <laravel-query-builder-geospatial-geoWithin>`
- :ref:`Intersecting a GeoJSON object <laravel-query-builder-geospatial-geoIntersects>`
- :ref:`Proximity data for nearby matches <laravel-query-builder-geospatial-geoNear>`

.. important::

   To perform GeoJSON queries in MongoDB, you must create either
   ``2d`` or ``2dsphere`` index on the collection. To learn more
   about these indexes, see :manual:`Geospatial Indexes </core/indexes/index-types/index-geospatial/>`
   in the {+server-docs-name+}.

To learn more about GeoJSON objects that MongoDB supports,
see :manual:`GeoJSON Objects </reference/geojson/>`
in the {+server-docs-name+}.

.. _laravel-query-builder-geospatial-near:

Near a Position Example
~~~~~~~~~~~~~~~~~~~~~~~

The following example shows how to use the ``near`` query operator
with the ``where()`` query builder method to match documents that
contain a location that is up to ``50`` meters from a GeoJSON Point

> object:

Reviewer 2: This should tell how to create a geo index.

Author: Will do -- I can link to the Schema Builder section "Create a Geospatial Index". However, I'll need to merge in the 4.1 branch to get that content. Is that ok or should I create a separate ticket to address this?
